### PR TITLE
Fix an issue with where clause with a dot character.

### PR DESCRIPTION
**Pull request by lfu**
```
 A SQL query with a condition value containing dot character causes join query based eager loading.
```

[Bug 1018392](https://bugzilla.redhat.com/show_bug.cgi?id=1018392)

Rails 3 has 2 approaches to avoid the N+1 queries problem. One involves creating a big join based query to pull in your associations, the other involves making a separate query per association.

When you call includes, rails decides which strategy to use. The default is to use the separate query approach unless it thinks you are referencing the columns from the associations in you conditions or order. Since that only works with the joins approach, it uses that instead.

Due to a bug in rails method tables_in_string, a SQL query that contains dot characters in the condition uses the join based query approach when the separate query approach is intended. 

The test cases in this commit are written to verify that the SQL query works as intended, specially with rails future release.

### CODE REVIEW

File: activerecord/lib/active_record/relation.rb [hunk 1]
```diff
@@ -529,9 +529,14 @@ def references_eager_loaded_tables?
 
     def tables_in_string(string)
       return [] if string.blank?
+
+      # a condition value like 'any_thing.v4' will be treated as a table by the scan pattern
+      # convert the condtion to 'any_thing-v4' before scan to avoid the join based eager loading
+      new = string.gsub(/('.*?')/) { $1.gsub(/\./, "-") }
```
Comment (Fryguy) [inline activerecord/lib/active_record/relation.rb]: Is it necessary to replace with a "-"?  Can we just remove it?

**Comment by lfu** [comment]
@Fryguy As long as the dot is removed, the where clause can be handled in any way. 
Done.

**Comment by Fryguy** [comment]
@lfu Sorry, I was thinking more along the lines of this, to remove as much overhead from this approach as possible:

``` ruby
new = string.gsub(/('.*?')/, "")
```

Even so, I'm concerned this isn't good enough to cover all of the cases.  I'm not sure how much detail is needed, but consider this where clause:  "name LIKE 'macy\'s.com'" (or however you escape a single quote).

**Comment by lfu** [comment]
@Fryguy We're repeating what have happened to rails :) Maybe upgrade to rails 4?

**Comment by Fryguy** [comment]
haha...I didn't realize that.  Got a link?

@jrafanie Can you review?  I'm ok with merging to get us over the hump.
@lfu I still think we should remove the block form of gsub.  This code will get run for every string condition, and may have a performance impact if we're not careful.

**Comment by lfu** [comment]
This [link](http://stackoverflow.com/questions/5199235/activerecord-query-changing-when-a-dot-period-is-in-condition-value) has detailed discussions.

**Comment by jrafanie** [comment]
:+1:

**Pull request merged**